lib/_stream_readable.js

```diff
@@ -1215,11 +1215,19 @@ function endReadableNT(state, stream) {
     state.endEmitted = true;
     stream.emit('end');
 
-    if (state.autoDestroy) {
+    if (stream.writable && stream.allowHalfOpen === false) {
+      stream.end();
```

Comment:
`Duplex` would do this in nextTick (https://github.com/nodejs/node/pull/31806/files#diff-4c74bb09dc4dabfff99ec3bc35181885L140) while `net.Socket` did it in same tick. Try using the `net.Socket` way as that is less breaking in terms of test.